### PR TITLE
Item Audit Trail

### PR DESCRIPTION
**Pull request by jimduff-usds**
This PR adds a new `item_lineage` table and populates it during the reports, batch, send, and download steps.

## Changes
- New `item_lineage` table and all the CRUD code the goes with it.  (Well, more like CR - no updating or deleting)
- Added a new organization called 'hub', to allow for thorough but fully automated testing of workflow.   
- To support that automation, changed the `./prime` cli to allow for `--input-fake` multiple `--target-counties` and multiple `--target-states`.
- To further support testing automation, added ability to generate fake data that round-robins thru a list of choices.
- See prime-router/docs/proposals/0003-lineage/additional-notes.md for implementation notes.

## Checklist
- [X] Tested locally?
- [ ] Added new dependencies?
- [X] Updated the docs?
- [X] Added a test?
- [X] Did you check for sensitive data, and remove any?
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

## Fixes 

- #426 

## To Be Done

- Not sure what to do about trackingIds for schemas that don't specify a trackingElement.  Should the original message_id trace all the way thru to redox and hl7 data, even though they drop that value from their data?
- Still have not cleaned up Task and TaskSource.

### CODE REVIEW

File: prime-router/src/test/kotlin/ReportTests.kt [hunk 1]
```diff
@@ -139,6 +139,140 @@ class ReportTests {
...
+    @Test
+    fun `test lineage insanity`() {
+        val schema = Schema(name = "one", topic = "test", elements = listOf(Element("a")), trackingElement = "a")
+        // each sublist is a row.
+        val report1 = Report(schema, listOf(listOf("bbb"), listOf("aaa"), listOf("aaa")), source = TestSource)
+        val metadata = Metadata(schema = schema)
+        val jurisdictionalFilter = metadata.findJurisdictionalFilter("matches") ?: fail("cannot find filter")
+
+        // split, merge, split, merge, copy, copy, then filter.
+        val reports1 = report1.split()
```
Comment (jimduff-usds) [inline prime-router/src/test/kotlin/ReportTests.kt:252]: Life is short - have fun.

File: prime-router/end-to-end-test.sh [hunk 2]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
+
+# Run a nice end to end test, covering all our formats, and all our transports,
+# and many different schemas.
+# 
+# This assumes a localhost prime/router is already running.
+# This assumes batch is on a 1-minute timer.
+#
+# This script is very ugly and repetitive.  Replace with kotlin.
+#
```
Comment (jimduff-usds) [inline prime-router/end-to-end-test.sh:10]: If you have the local prime-router docker-compose up and and running, you should be able to just run this script out of the box.   You have to eyeball the results, but I tried to make them very simple to read.   I plan to write this in Kotlin to make it much more malleable.

File: prime-router/src/main/resources/db/migration/V5__item_tables.sql [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+       UNION ALL   
+          SELECT IL.item_lineage_id, IL.child_report_id, IL.child_index
+          FROM item_lineage AS IL   
+          JOIN tmp ON (tmp.tmp_report_id = IL.parent_report_id
+          AND tmp.tmp_index = IL.parent_index)
+     )
+     SELECT tmp_lineage_id FROM tmp ORDER BY tmp_lineage_id;
+END; 
+$$  LANGUAGE PLPGSQL;    
+
```
Comment (jimduff-usds) [inline prime-router/src/main/resources/db/migration/V5__item_tables.sql:52]: Example calls to this stored proc can be seen in end-to-end-test.sh

File: prime-router/docs/openapi.yml [hunk 4]
```diff
@@ -127,12 +130,42 @@ components:
...
+        itemCount:
+          type: string
+          description: The number of items (rows, in a csv) due to be sent to this organization/service.
+      example:
+        organization: Brobdingnag Public Health Department
+        organization_id: brob-phd
+        service: elr
+        sending_at: 2021-02-11T14:55-05:00
+        itemCount: 17
+
```
Comment (jimduff-usds) [inline prime-router/docs/openapi.yml:171]: I had forgotten to update the swagger doc when I made the change to `destinations` (in a different PR)

File: prime-router/quick-test.sh [hunk 5]
```diff
@@ -225,17 +225,17 @@ then
   numitems=5
   echo Merge testing.  First, generate some fake STRAC data
    # Hack: put some unique strings in each one, so we can count lines.
-  text=$(./prime data --input-fake $numitems --input-schema strac/strac-covid-19 --output-dir $outputdir --target-county lilliput)
+  text=$(./prime data --input-fake $numitems --input-schema strac/strac-covid-19 --output-dir $outputdir --target-counties lilliput)
   parse_prime_output_for_filename "$text" $STRAC_FILE_SEARCH_STR
```
[the remark this answers is not in the document]
Comment (MauriceReeves-usds) [reply prime-router/quick-test.sh:229]: I like this!

File: prime-router/src/main/kotlin/FakeReport.kt [hunk 6]
```diff
@@ -265,10 +265,14 @@ class FakeReport(val metadata: Metadata) {
...
-        targetCounty: String? = null
+        targetStates: String? = null,
+        targetCounties: String? = null
     ): Report {
-        val rows = (0 until count).map { buildRow(schema, targetState, targetCounty) }.toList()
+        val counties = targetCounties?.split(",")
+        val states = targetStates?.split(",")
+        val rows = (0 until count).map {
+            buildRow(schema, roundRobinChoice(states), roundRobinChoice(counties))
+        }.toList()
```
[the remark this answers is not in the document]
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/FakeReport.kt:275]: This is great. Maybe in the future we could add a switch to do this kind of random vs truly random?

**Review by MauriceReeves-usds — APPROVED** [review]

File: prime-router/docs/proposals/0003-lineage/additional-notes.md [hunk 7]
```diff
@@ -23,17 +23,40 @@ One option would be to implement triggers that prevent updates except for a stat
...
+1. Success
+2. Failed to send
+3. Not sent, because it was already successfully sent in some previous attempt.
+
+Note that because of the 'all or nothing' assertion above, the retry code in RedoxTransport was modified to track item_lineage for all three of these statuses.
+
+**Main issue remaining**:  When a Report is read into BatchFunction whose schema has no trackingElement defined in the .schema, we lose the tracking_id column in the item_lineage table.  This is an artifact of how lineage tracking works in BatchFunction, where lineage is created in-memory based on data in the Report.kt objects.   If the parent report in the lineage has no trackingElement defined, then of course we can't fill in the item_lineage.tracking_id column.  (Currently most/all redox and hl7 schemas are affected by this, as they do not have trackingElements.)  A possible fix is to read in the parent's item_lineage from the database table.  However, that ID often does not exist in the child data - its not clear what the value is of tracking an ID that is not even in the data.  Further complexity with trackingElements will arise if we begin splitting/merging Items (see comment above).
+
+
```
Comment (RickHawesUSDS) [inline prime-router/docs/proposals/0003-lineage/additional-notes.md:62]: Perhaps we should require a tracking element in all schemas? If one is not valid, we should generate one?

File: prime-router/src/main/kotlin/Report.kt [hunk 8]
```diff
@@ -135,14 +147,17 @@ class Report {
...
-        bodyFormat: Format? = null
+        bodyFormat: Format? = null,
+        itemLineage: List<ItemLineage>? = null,
+        id: ReportId? = null // If constructing from blob storage, must pass in its UUID here.  Otherwise null.
     ) {
-        this.id = UUID.randomUUID()
+        this.id = id ?: UUID.randomUUID()
         this.schema = schema
         this.sources = sources
         this.createdDateTime = OffsetDateTime.now()
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Report.kt:157]: I would really like to get the sources out of this code.

**Review by RickHawesUSDS — COMMENTED** [review]
Look's really good. 

I guess the only thing I was hoping for was the removal of the report source stuff.

**Pull request merged**